### PR TITLE
C/C++ guide/tutorial

### PR DESCRIPTION
**Pull request by yyoncho**
Initial version of C/C++ language tutorial.

----

#

### CODE REVIEW

**Comment by sebastiansturm** [comment]
that is some amazing dedication to user-friendliness IMO. Seems comprehensive and to the point to me; I'm generally using more articles but I'm not a native speaker either so I might just be wrong. I'll add a few trivial remarks concerning grammar first, then try out your dap-mode tutorial. For the main part I'll need to set up a separate emacs config from scratch, will do that on the weekend

File: docs/blog/2020/CPP-guide.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+## Debugging
+
+Debugging in `lsp-mode` is implemented via [dap-mode](https://github.com/emacs-lsp/dap-mode) package. `dap-mode`
+front-end of the sister [Debug Adapter Protocol](https://microsoft.github.io/debug-adapter-protocol/). It is again client server
+like `Language Sever Protocol`. `dap-mode` provides all of the traditional
+debugger features - breakpoints(conditions, hit count, etc), threads, locals,
+
+### Installation
+
+`dap-mode` provides installation commands for most of the debug adapters. This
```
Comment (sebastiansturm) [inline docs/blog/2020/CPP-guide.md:216]:
```suggestion
`dap-mode` provides installation commands for most of the debug adapters. In this
```

File: docs/blog/2020/CPP-guide.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+tutorial we will use [vscode-cpptools](https://github.com/microsoft/vscode-cpptools) debug adapter. To install the adapter
```
Comment (sebastiansturm) [inline docs/blog/2020/CPP-guide.md:217]:
```suggestion
tutorial we will use the [vscode-cpptools](https://github.com/microsoft/vscode-cpptools) debug adapter. To install the adapter
```

File: docs/blog/2020/CPP-guide.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+do `M-x dap-cpptools-setup`.
+
+_Note:_ you should have `gbd` on path.
```
Comment (sebastiansturm) [inline docs/blog/2020/CPP-guide.md:220]:
```suggestion
_Note:_ you should have `gdb` on path.
```

**Review by sebastiansturm — COMMENTED** [review]
I'll look into the debugging part on the weekend, haven't set that up on my machine yet

File: docs/blog/2020/CPP-guide.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
+---
+disqus: emacs-lsp, C++/C
```
Comment (ericdallo) [inline docs/blog/2020/CPP-guide.md:2]: I think this should be our disqus username
```suggestion
disqus: emacs-lsp
```

File: docs/blog/2020/CPP-guide.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+---
+
+# Configuring `Emacs` as `C/C++` IDE and for `Emacs` core development
```
[the remark this answers is not in the document]
Comment (ericdallo) [reply docs/blog/2020/CPP-guide.md:5]: I'd reduce the title for some for attractive like:
```suggestion
# Configuring `Emacs` a `C/C++` IDE
```
And maybe mention about emacs core development somewhere bellow

File: docs/blog/2020/CPP-guide.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+In this guide, I will help you to setup an efficient working `C/C++`
+environment. The guide is mainly focused on `lsp-mode`/`dap-mode` and the
+features t. We will use `GNU Emacs` as a project in order to illustrate each of
```
Comment (ericdallo) [inline docs/blog/2020/CPP-guide.md:9]:
```suggestion
features that they have. We will use `GNU Emacs` as a project in order to illustrate each of
```

File: docs/blog/2020/CPP-guide.md [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+the steps to get `C/C++` project up and running.
+
+Some of the main features that we can expect from C++ IDE:
+
+* Code completion
+* Real-time syntax checking
+* Code navigation (references/definitions/implementations/symbol search)
+* Visual IDE debugger
+
```
Comment (ericdallo) [inline docs/blog/2020/CPP-guide.md:18]: Maybe mention other lsp main features
```suggestion

For all `lsp-mode` features, check [here](https://emacs-lsp.github.io/lsp-mode/page/main-features/)
```

**Review by ericdallo — COMMENTED** [review]
Well done! This is a really good post and I bet this will help a lot of new users.

There are lot of content of this post that I see that applies for other languages too, I wonder if we should have one post more focused on LSP/DAP features and another for clang.

File: docs/blog/2020/CPP-guide.md [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+- `lsp-rename` (`s-l r r`) - rename symbol/function at point.
+- `lsp-execute-code-action` (`s-l a a`) - `clangd` is able to auto-correct some of the errors.
+
+### Navigation
+`lsp-mode` has integration with `xref` core package and in addition it has
+
+- `xref-find-definitions`(`M-.` and `s-l g g`) - find definition(s) at point
+- `xref-find-references`(`s-l g r`) - find references to the symbol at point
+![References](images/images/references-helm.png "References helm")
+- `lsp-find-implementation`(`s-l g i`) - find implementation
```
[the remark this answers is not in the document]
Comment (yyoncho) [reply docs/blog/2020/CPP-guide.md:175]: Yes, I will delete the reference.

File: docs/blog/2020/CPP-guide.md [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+By default, `clangd` will assume your code is built as `clang` `some_file.cc`, and
+you’ll probably get spurious errors about missing `#included` files, etc. There
+are a couple of ways to fix this. You may read about them in [installation](https://clangd.llvm.org/installation.html "clangd installation")
+
+In order to generate `compile_commands.json` for `Emacs` itself I have found
+that [Bear](https://github.com/rizsotto/Bear) works fine.
+
+``` bash
+git clone https://github.com/rizsotto/Bear
```
Comment (kiennq) [inline docs/blog/2020/CPP-guide.md:55]: `Bear` can be installed via `apt` too `sudo apt install bear`

**Comment by danielmartin** [comment]
Thanks for the tutorial! I'll have a look at it this weekend.

**Comment by yyoncho** [comment]
> There are lot of content of this post that I see that applies for other languages too, I wonder if we should have one post more focused on LSP/DAP features and another for clang.

We can create lsp-mode generic section with things that apply to all servers. I am planning such an article but I want to sort out a 
 few issues before that. As we discussed in the past, the best option seems to be structuring the docs like vscode do - e. g. https://code.visualstudio.com/docs/nodejs/reactjs-tutorial

**Comment by tschuett** [comment]
Thanks for giving me lsp-mode! Please consider having a look at: https://github.com/clangd/clangd/issues/498

File: docs/blog/2020/CPP-guide.md [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+
+### Debugging
+
+Start debugging via `M-x dap-debug` and select `Debug Emacs` from the list.
+![select debug template](images/select-template.png "Select debug template").
+
+After pressing `RET` a fresh instance of `Emacs` will pop up. Then do `M-x
+helm-lsp-workspace-symbol` and type `message3`. Go to that function and place a
+breakpoint. You could do that by clicking in the fringe or by doing `M-x
+dap-breakpoint-toggle`. Switch to debugged instance and do
```
Comment (danielmartin) [inline docs/blog/2020/CPP-guide.md:275]:
```suggestion
dap-breakpoint-toggle`. Switch to the debugged instance and do
```

File: docs/blog/2020/CPP-guide.md [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+like `Language Sever Protocol`. `dap-mode` provides all of the traditional
+debugger features - breakpoints(conditions, hit count, etc), threads, locals,
+
+### Installation
+
+`dap-mode` provides installation commands for most of the debug adapters. In this
+tutorial we will use the [vscode-cpptools](https://github.com/microsoft/vscode-cpptools) debug adapter. To install the adapter
+do `M-x dap-cpptools-setup`.
+
+_Note:_ you should have `gdb` on path.
```
Comment (danielmartin) [inline docs/blog/2020/CPP-guide.md:229]: For macOS users, `lldb` is more popular than `gdb` (which is not available in the base system). As the debug adapter supports both I can expand the guide later to show how it works with `lldb`.

File: docs/blog/2020/CPP-guide.md [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+      ;; be more ide-ish
+      lsp-headerline-breadcrumb-enable t)
+
+(with-eval-after-load 'lsp-mode
+  (add-hook 'lsp-mode-hook #'lsp-enable-which-key-integration)
+  (require 'dap-cpptools)
+  (yas-global-mode))
+```
+
+# Feature overview.
```
Comment (danielmartin) [inline docs/blog/2020/CPP-guide.md:127]:
```suggestion
# Feature overview
```

File: docs/blog/2020/CPP-guide.md [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+when it is installed on the host machine and it is present on the path. In case
+`clangd` is not present on the path `clangd` can be installed following their
+[Getting Started](https://clangd.llvm.org/installation.html "Getting Started")
+instructions.
+
+_Note:_ for `Debian` based system use https://apt.llvm.org/
+
+## Project setup
+
+To understand your source code, `clangd` needs to know your build flags. (This is
```
[the remark this answers is not in the document]
Comment (yyoncho) [reply docs/blog/2020/CPP-guide.md:49]: I will apply that, but I realized that I have copied that section directly from clangd docs(but I already applied few comments on it).

File: docs/blog/2020/CPP-guide.md (hunk 11 above)
Comment (yyoncho) [reply docs/blog/2020/CPP-guide.md:229]: Yes. I was about to suggest someone using macOS(or other popular OS-es) to provide the macos specific steps.

File: docs/blog/2020/CPP-guide.md [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+
+Now, if everything was successful you will be asked to select a project root.
+
+![select project root](images/import-project.png "Select project root")
+
+_Note:_ the project root is needed by the language server in order to know where to
+start the project discovery from. Typically, this is the git repository root but since exceptions to this rule have caused us a lot of trouble in the past (monorepos come to mind), `lsp-mode` by default asks the user to manually confirm the project root the first time a project is opened.
+mono-repos it might be not. `lsp-mode` by default will ask the user to confirm
+the project root (automatic root selection has caused a lot of trouble for us in
+the past).
```
[the remark this answers is not in the document]
Comment (yyoncho) [reply docs/blog/2020/CPP-guide.md:143]: I prefer not to mention that option here. In the beginning, it was part of popular guides and a lot of users put it in their config which caused a lot of issues/complaints.

File: docs/blog/2020/CPP-guide.md [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+## Project setup
+
+To understand your source code, `clangd` needs to know your build flags. (This is
+just a fact of life in C++, source files are not self-contained).
+
+By default, `clangd` will assume your code is built as `clang` `some_file.cc`, and
+you’ll probably get spurious errors about missing `#included` files, etc. There
+are a couple of ways to fix this. You may read about them in [installation](https://clangd.llvm.org/installation.html "clangd installation")
+
+In order to generate `compile_commands.json` for `Emacs` itself I have found
```
[the remark this answers is not in the document]
Comment (yyoncho) [reply docs/blog/2020/CPP-guide.md:56]: It is mentioned in the linked Clangd resource - https://clangd.llvm.org/installation.html , I can copy their section?

We might consider providing also a "Hello world" example and also whatever you C++ guys think is relevant.

[the remark this answers is not in the document]
Comment (tschuett) [reply docs/blog/2020/CPP-guide.md:56]: https://sarcasm.github.io/notes/dev/compilation-database.html discusses various ways to get compile_commands.json files.

File: docs/blog/2020/CPP-guide.md (hunk 9 above)
Comment (yyoncho) [reply docs/blog/2020/CPP-guide.md:60]: thanks, updated.

File: mkdocs.yml [hunk 16]
```diff
@@ -100,6 +100,9 @@ nav:
       - page/gallery.md
   - Changelog:
       - page/CHANGELOG.md
+  - Tutorials:
+      - C/C++:
+          - 'Configuring Emacs a C/C++ IDE': tutorials/CPP-guide.md
```
Comment (ericdallo) [inline mkdocs.yml:105]: Will we need this kind of level of segregation? what about just:
```suggestion
  - Tutorials:
      - 'Configuring Emacs a C/C++ IDE': tutorials/CPP-guide.md
```

Comment (yyoncho) [reply mkdocs.yml]: I merged your proposal - we may revisit it later.

**Review by ericdallo — APPROVED** [review]
:clap:

**Pull request merged**